File: user-service/src/main/java/rs/raf/user_service/service/VerificationRequestService.java
package rs.raf.user_service.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import rs.raf.user_service.client.BankClient;
import rs.raf.user_service.domain.dto.CardRequestDto;
import rs.raf.user_service.domain.dto.CreateVerificationRequestDto;
import rs.raf.user_service.domain.entity.VerificationRequest;
import rs.raf.user_service.domain.enums.VerificationStatus;
import rs.raf.user_service.domain.enums.VerificationType;
import rs.raf.user_service.exceptions.RejectNonPendingRequestException;
import rs.raf.user_service.exceptions.VerificationNotFoundException;
import rs.raf.user_service.repository.VerificationRequestRepository;
import rs.raf.user_service.utils.JwtTokenUtil;

import javax.persistence.EntityNotFoundException;
import java.time.LocalDateTime;
import java.util.List;

@AllArgsConstructor
@Service
public class VerificationRequestService {
    private final VerificationRequestRepository verificationRequestRepository;
    private final BankClient bankClient;
    private final JwtTokenUtil jwtTokenUtil;
    private ObjectMapper objectMapper;


    public void createVerificationRequest(CreateVerificationRequestDto createVerificationRequestDto) {
        VerificationRequest request = VerificationRequest.builder()
                .userId(createVerificationRequestDto.getUserId())
                .targetId(createVerificationRequestDto.getTargetId())
                .status(VerificationStatus.PENDING)
                .verificationType(createVerificationRequestDto.getVerificationType())
                .expirationTime(LocalDateTime.now().plusMinutes(5))
                .details(createVerificationRequestDto.getDetails())
                .build();

        verificationRequestRepository.save(request);
    }


    public List<VerificationRequest> getActiveRequests(Long userId) {
        return verificationRequestRepository.findActiveRequests(userId);
    }

    public List<VerificationRequest> getRequestHistory(Long userId) {
        return verificationRequestRepository.findInactiveRequests(userId);
    }

    public boolean updateRequestStatus(Long requestId, VerificationStatus status) {
        return verificationRequestRepository.findById(requestId).map(request -> {
            request.setStatus(status);
            verificationRequestRepository.save(request);
            return true;
        }).orElse(false);
    }

    public boolean processApproval(Long requestId, String authHeader) {
        Long clientIdFromToken = jwtTokenUtil.getUserIdFromAuthHeader(authHeader);

        VerificationRequest request = verificationRequestRepository.findActiveRequest(requestId, clientIdFromToken)
                .orElseThrow(() -> new IllegalStateException("Verification request not found"));

        request.setStatus(VerificationStatus.APPROVED);
        verificationRequestRepository.save(request);

        switch (request.getVerificationType()) {
            case CHANGE_LIMIT -> bankClient.changeAccountLimit(request.getTargetId());//radi
            case PAYMENT -> bankClient.confirmPayment(request.getTargetId());//radi
            case TRANSFER -> bankClient.confirmTransfer(request.getTargetId());
            case CARD_REQUEST -> bankClient.approveCardRequest(request.getTargetId());//radi
        }

        return true;
    }


    public void denyVerificationRequest(Long requestId, String authHeader) {
        Long clientIdFromToken = jwtTokenUtil.getUserIdFromAuthHeader(authHeader);

        VerificationRequest request = verificationRequestRepository.findActiveRequest(requestId, clientIdFromToken)
                .orElseThrow(() -> new VerificationNotFoundException(requestId));

        if(!request.getStatus().equals(VerificationStatus.PENDING))
            throw new RejectNonPendingRequestException();

        request.setStatus(VerificationStatus.DENIED);
        verificationRequestRepository.save(request);

        switch (request.getVerificationType()) {
            case CHANGE_LIMIT -> bankClient.rejectChangeAccountLimit(request.getTargetId());
            case PAYMENT -> bankClient.rejectConfirmPayment(request.getTargetId());
            case TRANSFER -> bankClient.rejectConfirmTransfer(request.getTargetId());
            case CARD_REQUEST -> bankClient.rejectApproveCardRequest(request.getTargetId());
        }
    }

    public boolean calledFromMobile(String userAgent) {
        return userAgent != null && userAgent.equals("MobileApp/1.0");
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/service/VerificationRequestService.kt
package pack.userservicekotlin.service

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import org.springframework.stereotype.Service
import pack.userservicekotlin.arrow.VerificationServiceError
import pack.userservicekotlin.domain.dto.verification.CreateVerificationRequestDto
import pack.userservicekotlin.domain.entities.VerificationRequest
import pack.userservicekotlin.domain.enums.VerificationStatus
import pack.userservicekotlin.domain.enums.VerificationType
import pack.userservicekotlin.external.BankClient
import pack.userservicekotlin.repository.VerificationRequestRepository
import pack.userservicekotlin.utils.JwtTokenUtil
import java.time.LocalDateTime

@Service
class VerificationRequestService(
    private val verificationRequestRepository: VerificationRequestRepository,
    private val bankClient: BankClient,
    private val jwtTokenUtil: JwtTokenUtil,
) {
    fun createVerificationRequest(dto: CreateVerificationRequestDto): Either<VerificationServiceError, Unit> {
        val request =
            VerificationRequest(
                userId = dto.userId,
                targetId = dto.targetId,
                status = VerificationStatus.PENDING,
                verificationType = dto.verificationType,
                expirationTime = LocalDateTime.now().plusMinutes(5),
                details = dto.details,
            )

        return try {
            verificationRequestRepository.save(request)
            Unit.right()
        } catch (e: Exception) {
            VerificationServiceError.DatabaseError("Failed to save verification request: ${e.message}").left()
        }
    }

    fun getActiveRequests(userId: Long): Either<VerificationServiceError, List<VerificationRequest>> =
        try {
            verificationRequestRepository.findActiveRequests(userId).right()
        } catch (e: Exception) {
            VerificationServiceError.DatabaseError("Failed to retrieve active requests: ${e.message}").left()
        }

    fun getRequestHistory(userId: Long): Either<VerificationServiceError, List<VerificationRequest>> =
        try {
            verificationRequestRepository.findInactiveRequests(userId).right()
        } catch (e: Exception) {
            VerificationServiceError.DatabaseError("Failed to retrieve request history: ${e.message}").left()
        }

    fun updateRequestStatus(
        requestId: Long,
        status: VerificationStatus,
    ): Either<VerificationServiceError, Boolean> {
        val request =
            verificationRequestRepository.findById(requestId).orElse(null)
                ?: return VerificationServiceError.RequestNotFound(requestId).left()

        return try {
            request.status = status
            verificationRequestRepository.save(request)
            true.right()
        } catch (e: Exception) {
            VerificationServiceError.DatabaseError("Failed to update request status: ${e.message}").left()
        }
    }

    fun processApproval(
        requestId: Long,
        authHeader: String,
    ): Either<VerificationServiceError, Boolean> {
        val clientIdFromToken = jwtTokenUtil.getUserIdFromAuthHeader(authHeader)

        val request =
            verificationRequestRepository.findActiveRequest(requestId, clientIdFromToken)
                ?: return VerificationServiceError.RequestNotFound(requestId).left()

        request.status = VerificationStatus.APPROVED
        verificationRequestRepository.save(request)

        return try {
            when (request.verificationType) {
                VerificationType.CHANGE_LIMIT -> {
                    bankClient.changeAccountLimit(request.targetId)
                    true.right()
                }
                VerificationType.PAYMENT -> {
                    bankClient.confirmPayment(request.targetId)
                    true.right()
                }
                VerificationType.TRANSFER -> {
                    bankClient.confirmTransfer(request.targetId)
                    true.right()
                }
                VerificationType.CARD_REQUEST -> {
                    bankClient.approveCardRequest(request.targetId)
                    true.right()
                }
                VerificationType.LOGIN -> TODO()
                VerificationType.LOAN -> TODO()
                null -> TODO()
            }
        } catch (e: Exception) {
            VerificationServiceError.BankServiceError("Bank service error: ${e.message}").left()
        }
    }

    fun denyVerificationRequest(
        requestId: Long,
        authHeader: String,
    ): Either<VerificationServiceError, Unit> {
        val clientIdFromToken = jwtTokenUtil.getUserIdFromAuthHeader(authHeader)

        val request =
            verificationRequestRepository.findActiveRequest(requestId, clientIdFromToken)
                ?: return VerificationServiceError.RequestNotFound(requestId).left()

        if (request.status != VerificationStatus.PENDING) {
            return VerificationServiceError.InvalidRequestStatus("Cannot deny a non-pending request").left()
        }

        request.status = VerificationStatus.DENIED
        verificationRequestRepository.save(request)

        return try {
            when (request.verificationType) {
                VerificationType.CHANGE_LIMIT -> {
                    bankClient.rejectChangeAccountLimit(request.targetId)
                    Unit.right()
                }
                VerificationType.PAYMENT -> {
                    bankClient.rejectConfirmPayment(request.targetId)
                    Unit.right()
                }
                VerificationType.TRANSFER -> {
                    bankClient.rejectConfirmTransfer(request.targetId)
                    Unit.right()
                }
                VerificationType.CARD_REQUEST -> {
                    bankClient.rejectApproveCardRequest(request.targetId)
                    Unit.right()
                }
                VerificationType.LOGIN -> TODO()
                VerificationType.LOAN -> TODO()
                null -> TODO()
            }
        } catch (e: Exception) {
            VerificationServiceError.BankServiceError("Bank service error: ${e.message}").left()
        }
    }

    fun calledFromMobile(userAgent: String?): Boolean = userAgent == "MobileApp/1.0"
}
